### PR TITLE
feat: support upgrade command

### LINKED ISSUE #752: Support datree upgrade command
**Describe the solution you'd like**

Add new `upgrade` command to install the latest datree version. 

- The command should support only macOS & Linux.
- The command should check if the user installed Datree with brew and if so, the command should fail and return a message “Looks like you installed Datree via brew, to upgrade datree run - brew upgrade datree”
- The command should not support pinned or user-defined versions.

**Requirements**
Golang mid-level.

**Reference:** [https://github.com/cli/cli/blob/trunk/pkg/cmd/extension/command.go](https://github.com/cli/cli/blob/trunk/pkg/cmd/extension/command.go)

### PR DESCRIPTION
**Pull request by amustaque97**
closes #752 

Here is the quick demo screencast — `datree upgrade` command

https://user-images.githubusercontent.com/21127788/195978978-ccd97fff-969f-45db-94b0-28b992df238f.mov

### CODE REVIEW

**Comment by shmu3l** [comment]
@amustaque97 It looks great! I'll check it out and promote it further in the next few days

**Comment by shmu3l** [comment]
Hey @amustaque97!

i reviewed your PR, regarding the logic for upgrading, we already have it in [`install.sh`](https://github.com/datreeio/datree/blob/main/install.sh), so no need to write it again, you can use the existing script.

WDYT?

**Comment by amustaque97** [comment]
> Hey @amustaque97!
> 
> i reviewed your PR, regarding the logic for upgrading, we already have it in [`install.sh`](https://github.com/datreeio/datree/blob/main/install.sh), so no need to write it again, you can use the existing script.
> 
> WDYT?

I would say, if we have more plans/ideas then we use golang code only because we can extend this functionality more easily

else

bash code is also fine. 


It boils down to personal preference because I personally prefer go code rather than a bash script 😛  because of silly mistakes and hard time in debugging. I'm fine either way. Just that it should solve our purpose.

Looking forward to hearing from you.

**Comment by myishay** [comment]
I think the main concern here is that we don't want to maintain two pieces of code with basically the same logic.
The bash script already exists and has to stay for the initial installation, so even if it might be easier to debug and read/write golang than bash script, it will make us maintain the same logic in 2 different code blocks.

@amustaque97 @shmu3l WDYT?

**Comment by amustaque97** [comment]
Hey @shmu3l and @myishay I have made the code changes and now I'm relying on the `install.sh` script content. I tested it in my local. It solves the purpose. Please review my PR.

File: cmd/upgrade/main.go
Comment (shalev007) [inline cmd/upgrade/main.go]: This is small, but let's try to keep the variable naming to camel case.
```suggestion
						errMsg := fmt.Sprintf("Failed to upgrade datree to the latest version, error encountered %s\n", err)
```

**Review by shalev007 — CHANGES_REQUESTED** [review]
Looks great!
left a small comment, I'll approve after this minor fix

File: pkg/upgradeManager/upgradeManager.go
Comment (royhadad) [inline pkg/upgradeManager/upgradeManager.go]: This is not bad, however, we should fetch the script from "https://get.datree.io", since this is the actual production install script

File: pkg/upgradeManager/upgradeManager.go
Comment (royhadad) [inline pkg/upgradeManager/upgradeManager.go]: This is the installation one-liner from the datree docs:
https://hub.datree.io/#quickstart

This seems a bit cumbersome to me, we can just execute the one-liner using Go's standard library, maybe something like this? What do you think?

```
out, err := exec.Command("bash", "curl https://get.datree.io | /bin/bash").Output()
```

File: cmd/upgrade/main.go
Comment (royhadad) [inline cmd/upgrade/main.go]: This should be checked in the "args" function, check out this example:
https://github.com/datreeio/datree/blob/4cb0e38c252a102ec8bb0b6010ebe2633dc0a19c/cmd/test/main.go#L200

Actually, even better, this check can be deleted completely! Cobra checks it by default

Screen Shot 2022-10-23 at 22 17 28

File: cmd/upgrade/main.go
Comment (royhadad) [inline cmd/upgrade/main.go:21]: Because this is an I/O dependency, adding it to the `UpgradeCommandContext` and injecting it from the outside in cmd/root.go makes more sense to me.
It will allow for easier testing

File: cmd/upgrade/main.go
Comment (royhadad) [inline cmd/upgrade/main.go]:
```suggestion
					ctx.Printer.PrintMessage("Datree upgraded successfully\n", "green")
```

**Review by royhadad — CHANGES_REQUESTED** [review]

File: pkg/upgradeManager/upgradeManager.go
Comment (amustaque97) [reply pkg/upgradeManager/upgradeManager.go]: Thanks for the heads-up. I think "https://get.datree.io" is short and easy to implement.

File: cmd/upgrade/main.go
Comment (amustaque97) [reply cmd/upgrade/main.go]: Thank you for the comment. I didn't know that 😅 . I removed the condition.

File: pkg/upgradeManager/upgradeManager.go
Comment (amustaque97) [reply pkg/upgradeManager/upgradeManager.go]: thanks for the review comment. After reading the issue description I thought we want to replicate a similar coding style but now I understood we just want implementation. I got confused 😕 

I have made the required changes. Code is much simpler and short.

**Comment by amustaque97** [comment]
@royhadad, thank you for the brief PR review. I have made the required changes. New code changes work fine, tested locally. Sharing a screenshot for reference. (In the screenshot I uninstalled datree intentionally to show downloading script works fine and similarly, it will upgrade existing datree using shell script)

Screenshot 2022-10-24 at 1 40 30 AM

File: cmd/upgrade/main.go
Comment (royhadad) [inline cmd/upgrade/main.go:27]: seems like this is a duplicated if statement :)

File: pkg/upgradeManager/upgradeManager.go
Comment (royhadad) [inline pkg/upgradeManager/upgradeManager.go]: This seems to work as well, replacing the entire function with 3 lines of code.

```
oneLineInstallationCommand := exec.Command("bash", "-c", "curl https://get.datree.io | /bin/bash")
oneLineInstallationCommand.Stdout = os.Stdout
return oneLineInstallationCommand.Run()
```

Is there any benefit to using the current 20 lines of code instead of the suggested 3? I'm asking out of curiosity

**Review by royhadad — CHANGES_REQUESTED** [review]
last 2 comments :)

File: cmd/upgrade/main.go
Comment (amustaque97) [reply cmd/upgrade/main.go:27]: I don't think so, In the issue, it is mentioned that we need to check if datree is installed using brew or not. In `install.sh` I don't see that logic with the brew command. Correct me If I'm wrong

File: pkg/upgradeManager/upgradeManager.go
Comment (amustaque97) [reply pkg/upgradeManager/upgradeManager.go]: I have barely seen `-c` option and if you're confident about the one-liner. I would love to see it happening. I have changed the code with the given code block. 👍🏻

**Review by shalev007 — DISMISSED** [review]

File: cmd/upgrade/main.go
Comment (royhadad) [reply cmd/upgrade/main.go:27]: look at line 27 and line 30, the same `if` statement is written twice :)

**Review by royhadad — CHANGES_REQUESTED** [review]

File: cmd/upgrade/main.go
Comment (amustaque97) [reply cmd/upgrade/main.go:27]: oops my mistake, sorry @royhadad, I didn't notice that I messed it up somehow when I took a pull from the remote branch when I committed the suggested changes from your on Github.

I hope everything is fine now.

**Comment by royhadad** [comment]
@amustaque97 since you got reviews from several people already, I didn't want to bother you for to 4th time and added a small fix
1. check for windows OS
2. print out stderr as well

hoping that is ok :)

**Review by royhadad — APPROVED** [review]

**Comment by royhadad** [comment]
@amustaque97 Thank you for the contribution, awesome work!! 🚀🚀🚀

**Pull request merged**
